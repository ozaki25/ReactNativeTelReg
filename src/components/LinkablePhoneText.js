import React from 'react';
import { StyleSheet, Text } from 'react-native';
import PropTypes from 'prop-types';
import PhoneText from './PhoneText';

// 0xxx-xxx-xxx or 0xx-xxx-xxxx or 0x-xxxx-xxxx
const reg = /0[0-9]{3}-[0-9]{3}-[0-9]{3}|0[0-9]{2}-[0-9]{3}-[0-9]{4}|0[0-9]{1}-[0-9]{4}-[0-9]{4}/g;

// ex. message = 'お問い合わせは03-0000-0000もしくは0120-000-000へお願いします'
const replacePhoneNumber = message => {
  // ex. phoneNumbers = [03-0000-0000, 0120-000-000]
  const phoneNumbers = message.match(reg);
  if (!phoneNumbers) return message;

  // ex. message.split(reg) = ['お問い合わせは', 'もしくは', 'へお願いします']
  return message.split(reg).reduce((result, text, i) => {
    // ex. ループ一周目 result = [], text = 'お問い合わせは', i = 0
    // ex. ループ二周目 result = ['お問い合わせは', <PhoneTtext phoneNumber="03-0000-0000" />], text = 'もしくは', i = 1
    // ex. ループ三周目 result = ['お問い合わせは', <PhoneTtext phoneNumber="03-0000-0000" />, 'もしくは', <PhoneTtext phoneNumber="0120-000-000" />], text = 'へお願いします', i = 2
    return [...result, text, phoneNumbers[i] && <PhoneText key={i} phoneNumber={phoneNumbers[i]} />];
  }, []);
};

/*
コメントない方が見やすいかも
const replacePhoneNumber = message => {
  const phoneNumbers = message.match(reg);
  if (!phoneNumbers) return message;

  return message.split(reg).reduce((result, text, i) => {
    return [...result, text, phoneNumbers[i] && <PhoneText key={i} phoneNumber={phoneNumbers[i]} />];
  }, []);
};
*/

const styles = StyleSheet.create({
  text: {
    color: '#bbb',
    fontSize: 16,
  },
});

const LinkablePhoneText = ({ children }) => (
  <Text style={styles.text}>{replacePhoneNumber(children)}</Text>
);

LinkablePhoneText.propTypes = {
  children: PropTypes.string.isRequired,
};

export default LinkablePhoneText;
