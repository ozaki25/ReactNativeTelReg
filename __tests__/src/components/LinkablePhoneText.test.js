import 'react-native';
import React from 'react';
import LinkablePhoneText from '../../../src/components/LinkablePhoneText';
import renderer from 'react-test-renderer';

describe('#LinkablePhoneText', () => {
  describe('電話番号だけの場合', () => {
    describe('電話番号が一つの場合', () => {
      it('0xxx-xxx-xxx形式の文字列がリンク化されること', () => {
        const children = '0120-123-123';
        const tree = renderer.create(<LinkablePhoneText>{children}</LinkablePhoneText>).toJSON();
        expect(tree).toMatchSnapshot();
      });
      it('0xx-xxx-xxxx形式の文字列がリンク化されること', () => {
        const children = '045-123-1234';
        const tree = renderer.create(<LinkablePhoneText>{children}</LinkablePhoneText>).toJSON();
        expect(tree).toMatchSnapshot();
      });
      it('0x-xxxx-xxxx形式の文字列がリンク化されること', () => {
        const children = '03-1234-1234';
        const tree = renderer.create(<LinkablePhoneText>{children}</LinkablePhoneText>).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
    describe('電話番号が複数の場合', () => {
      it('それぞれの電話番号がリンク化されること', () => {
        const children = '0120-123-123045-123-123403-1234-1234';
        const tree = renderer.create(<LinkablePhoneText>{children}</LinkablePhoneText>).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
  describe('電話番号と文字列を含む場合', () => {
    describe('電話番号が一つの場合', () => {
      it('電話番号から始まる場合、電話番号の文字列がリンク化されること', () => {
        const children = '03-1234-1234にご連絡ください';
        const tree = renderer.create(<LinkablePhoneText>{children}</LinkablePhoneText>).toJSON();
        expect(tree).toMatchSnapshot();
      });
      it('電話番号で終わる場合、電話番号の文字列がリンク化されること', () => {
        const children = '次の番号にご連絡ください03-1234-1234';
        const tree = renderer.create(<LinkablePhoneText>{children}</LinkablePhoneText>).toJSON();
        expect(tree).toMatchSnapshot();
      });
      it('文の途中に電話番号がある場合、電話番号の文字列がリンク化されること', () => {
        const children = 'お問い合わせは03-1234-1234へお願いします';
        const tree = renderer.create(<LinkablePhoneText>{children}</LinkablePhoneText>).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
    describe('電話番号が複数の場合', () => {
      it('それぞれの電話番号がリンク化されること', () => {
        const children = 'お問い合わせは0120-123-123または045-123-1234\n03-1234-1234へお願いします';
        const tree = renderer.create(<LinkablePhoneText>{children}</LinkablePhoneText>).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
  describe('文字列だけの場合', () => {
    it('改行なしの場合、そのままの文字列が表示されること', () => {
      const children = '窓口までお越し下さい';
      const tree = renderer.create(<LinkablePhoneText>{children}</LinkablePhoneText>).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('改行ありの場合、そのままの文字列が表示されること', () => {
      const children = '恐れ入りますが、\n窓口までお越し下さい';
      const tree = renderer.create(<LinkablePhoneText>{children}</LinkablePhoneText>).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
