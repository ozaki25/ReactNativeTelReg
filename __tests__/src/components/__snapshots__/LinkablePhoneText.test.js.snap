// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#LinkablePhoneText 文字列だけの場合 改行ありの場合、そのままの文字列が表示されること 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#bbb",
      "fontSize": 16,
    }
  }
>
  恐れ入りますが、
窓口までお越し下さい
</Text>
`;

exports[`#LinkablePhoneText 文字列だけの場合 改行なしの場合、そのままの文字列が表示されること 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#bbb",
      "fontSize": 16,
    }
  }
>
  窓口までお越し下さい
</Text>
`;

exports[`#LinkablePhoneText 電話番号だけの場合 電話番号が一つの場合 0x-xxxx-xxxx形式の文字列がリンク化されること 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#bbb",
      "fontSize": 16,
    }
  }
>
  
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    03-1234-1234
  </Text>
  
</Text>
`;

exports[`#LinkablePhoneText 電話番号だけの場合 電話番号が一つの場合 0xx-xxx-xxxx形式の文字列がリンク化されること 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#bbb",
      "fontSize": 16,
    }
  }
>
  
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    045-123-1234
  </Text>
  
</Text>
`;

exports[`#LinkablePhoneText 電話番号だけの場合 電話番号が一つの場合 0xxx-xxx-xxx形式の文字列がリンク化されること 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#bbb",
      "fontSize": 16,
    }
  }
>
  
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    0120-123-123
  </Text>
  
</Text>
`;

exports[`#LinkablePhoneText 電話番号だけの場合 電話番号が複数の場合 それぞれの電話番号がリンク化されること 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#bbb",
      "fontSize": 16,
    }
  }
>
  
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    0120-123-123
  </Text>
  
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    045-123-1234
  </Text>
  
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    03-1234-1234
  </Text>
  
</Text>
`;

exports[`#LinkablePhoneText 電話番号と文字列を含む場合 電話番号が一つの場合 文の途中に電話番号がある場合、電話番号の文字列がリンク化されること 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#bbb",
      "fontSize": 16,
    }
  }
>
  お問い合わせは
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    03-1234-1234
  </Text>
  へお願いします
</Text>
`;

exports[`#LinkablePhoneText 電話番号と文字列を含む場合 電話番号が一つの場合 電話番号から始まる場合、電話番号の文字列がリンク化されること 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#bbb",
      "fontSize": 16,
    }
  }
>
  
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    03-1234-1234
  </Text>
  にご連絡ください
</Text>
`;

exports[`#LinkablePhoneText 電話番号と文字列を含む場合 電話番号が一つの場合 電話番号で終わる場合、電話番号の文字列がリンク化されること 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#bbb",
      "fontSize": 16,
    }
  }
>
  次の番号にご連絡ください
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    03-1234-1234
  </Text>
  
</Text>
`;

exports[`#LinkablePhoneText 電話番号と文字列を含む場合 電話番号が複数の場合 それぞれの電話番号がリンク化されること 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#bbb",
      "fontSize": 16,
    }
  }
>
  お問い合わせは
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    0120-123-123
  </Text>
  または
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    045-123-1234
  </Text>
  

  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": 16,
        "textDecorationLine": "underline",
      }
    }
  >
    03-1234-1234
  </Text>
  へお願いします
</Text>
`;
